src/ray/core_worker/core_worker.cc

```diff
@@ -2169,29 +2172,41 @@ Status CoreWorker::PushError(const JobID &job_id,
   return local_raylet_client_->PushError(job_id, type, error_message, timestamp);
 }
 
-json CoreWorker::OverrideRuntimeEnv(json &child, const std::shared_ptr<json> parent) {
+json CoreWorker::OverrideRuntimeEnv(const json &child, std::shared_ptr<json> parent) {
   // By default, the child runtime env inherits non-specified options from the
   // parent. There is one exception to this:
   //     - The env_vars dictionaries are merged, so environment variables
   //       not specified by the child are still inherited from the parent.
   json result_runtime_env = *parent;
-  for (json::iterator it = child.begin(); it != child.end(); ++it) {
+  for (auto it = child.cbegin(); it != child.cend(); ++it) {
     if (it.key() == "env_vars" && result_runtime_env.contains("env_vars")) {
       json env_vars = it.value();
       json merged_env_vars = result_runtime_env["env_vars"];
       for (json::iterator nit = env_vars.begin(); nit != env_vars.end(); ++nit) {
         merged_env_vars[nit.key()] = nit.value();
       }
-      result_runtime_env["env_vars"] = merged_env_vars;
+      result_runtime_env["env_vars"] = std::move(merged_env_vars);
     } else {
       result_runtime_env[it.key()] = it.value();
     }
   }
   return result_runtime_env;
 }
 
+// TODO(hjiang): Current implementation is not the most ideal version, since it acquires a
+// global lock for all operations; it's acceptable for now since no heavy-lifted operation
+// is involved (considering the overall scheduling overhead is single-digit millisecond
+// magnitude). But a better solution is LRU cache native providing a native support for
+// sharding and `GetOrCreate` API.
 std::shared_ptr<rpc::RuntimeEnvInfo> CoreWorker::OverrideTaskOrActorRuntimeEnvInfo(
     const std::string &serialized_runtime_env_info) const {
+  std::lock_guard lck(job_runtime_env_serialization_mutex_);
+  if (auto cached_runtime_env_info =
+          runtime_env_json_serialization_cache_.Get(serialized_runtime_env_info);
```

Comment:
I only see this Get, but no Put. Where did you Put()?

Also, on job change we need to reset the cache and I don't see that either. It should be fine because today a core worker is hard wired to a job, and never changes jobs. But if you find it's easy to add that, feel free to add.

Reply:
> I only see this Get, but no Put. Where did you Put()?

.... fixed.

Reply:
> Also, on job change we need to reset the cache and I don't see that either. It should be fine because today a core worker is hard wired to a job, and never changes jobs. But if you find it's easy to add that, feel free to add.

I'm not familiar with this part, I left a TODO and for next PR, will discuss with you in person.